### ISSUE TITLE
Syntax for COUNT(), AVG() and SUM()

### ISSUE THREAD

**Issue by adnan-kamili** [typeorm/typeorm#673]
I didn't find any docs for operations like count(), avg(), sum(). Something like:

```javascript
const productCount = await this.productRepository.count({ accountId });
const itemCount = await this.productRepository.find({accountId}).sum("items");
```

**Comment by pleerock** [comment]
use query builder to build any sql you need and get raw results. Something like:

```ts
let [sum] = await context.locations
            .createQueryBuilder("location")
            .select("SUM(location.something)", "sum")
            .getRawOne(); 
```

**Comment by adnan-kamili** [comment]
Thanks! It would be nice if there were functions for the same as these operations are quite common and other orms support them out of the box.

**Comment by pleerock** [comment]
Maybe. In the future.

**Comment by greg-murray-volusion** [comment]
> use query builder to build any sql you need and get raw results. Something like:
> 
> ```ts
> let [sum] = await context.locations
>             .createQueryBuilder("location")
>             .select("SUM(location.something)", "sum")
>             .getRawOne(); 
> ```

I think this one should be 

```typescript
let { sum } = await ...
```

**Comment by syuilo** [comment]
Yes, as @greg-murray-volusion said, `let [sum] =` is not working.
Use `let { sum } =` instead.

**Comment by zaksesti-ns8** [comment]
Is this still true?  I see SelectQueryBuilder has a `getCount()` method.

**Comment by AlexWehe** [comment]
> Maybe. In the future.

https://youtu.be/0UkYwZlTLxE

**Comment by SimantoR** [comment]
> Maybe. In the future.

Future is now. I understand that most people don't have the time and it only makes sense that people using the library should be contributing to the source-code the most. No one does things for no value what-so-ever, which is apparent.

**Comment by iamchathu** [comment]
Is there any update on this?

**Comment by luka-rmar** [comment]
Any updates for this functionality?

**Comment by italodeandra** [comment]
If anyone is looking for on how to get the list and the count (the count doesn't follow the `take`/`skip`), you can find an example here (the only place I could find in the official website):

[https://typeorm.io/#undefined/loading-from-the-database - This link doesn't work. Just a reference](https://typeorm.io/#undefined/loading-from-the-database)

But as I just notice the link doesn't work. You can find on the **Loading from the database** section on https://typeorm.io/#/

And here is the code:

```
let [allPhotos, photosCount] = await photoRepository.findAndCount();
console.log("All photos: ", allPhotos);
console.log("Photos count: ", photosCount);
```

It still works with **where** filter, and the `count` ignores the `take`/`skip` so you get the count without your pagination.

**Comment by AtaruOhto** [comment]
In our project, we needed to calculate an average and truncate in transaction flow (MySQL). I managed to implement this feature with TypeORM as following:

```
// Transaction
connection.transaction(async (transactionalEntityManager) => {
      ...
      const reviewsScoreAverage = await transactionalEntityManager
        .createQueryBuilder()
        .select("TRUNCATE(AVG(score), 1)", "score")
        .from(Review, "reviews")
        .getRawOne();
      ...
  });
```

```
// SQL output
// SELECT TRUNCATE(AVG(score),1) AS `score` FROM `reviews` `reviews`
```

**Comment by ahmadalfy** [comment]
The current version of typeorm now supports `count`. This will work

```javascript
const productCount = await this.productRepository.count({ accountId });
```

**Comment by qiwang97** [comment]
any news about `SUM` and `AVG` like `count ` syntax ?

**Comment by chinmay-mfl** [comment]
> The current version of typeorm now supports `count`. This will work
> 
> ```js
> const productCount = await this.productRepository.count({ accountId });
> ```

I think this should be:
```js 
const productCount = await this.productRepository.count({ where: {userId: accountId} });
```

**Comment by EliasTouil** [comment]
Maybe this is not the answer you want to hear, but need to hear :
I've been using TypeORM for a year now and it's hard to find examples sometimes. Oftentimes I think it's because of my lack of knowledge in SQL and POSTGRESQL, I don't know what I'm looking for.
When I'm unable to find what I need, a quick workaround is to use the raw SQL queries, as there are a lot of good references for these technologies. 

I understand this is not the best, or what you may be looking for right now, but I hope these examples help you move to your next ticket. 

### What is the average session duration?
```typescript
@Resolver(UserSession)
export class UserSessionResolver {
  @Query(() => Number)
  async getSessionDurationAvg() {
    const result = await getManager().query(`
      SELECT AVG(duration) FROM user_session AS avg;
    `);

    return Math.floor(result[0].avg) || 0;
  }
}
```

### What is the count of article views which have a distinct user_id 
```typescript
@Resolver(View)
export class ViewResolver {
  @UseMiddleware(allowRoles([RoleName.ADMIN, RoleName.STAFF]))
  @Query(() => Int)
  async getBillableSessionsCount(): Promise<number> {
    const result = await getManager().query(`
      SELECT COUNT(*) FROM(SELECT DISTINCT "user_id" FROM "view") AS count;
    `);
    return result[0].count || 0;
  }
}
```

**Comment by iamnotstatic** [comment]
any news about ```SUM``` and ```AVG``` like count syntax ?

**Comment by lehaiquantb** [comment]
Any new Update about ```AVG```

**Comment by malinovaya0iriska** [comment]
> > The current version of typeorm now supports `count`. This will work
> > ```js
> > const productCount = await this.productRepository.count({ accountId });
> > ```
> 
> I think this should be:
> 
> ```js
> const productCount = await this.productRepository.count({ where: {userId: accountId} });
> ```

Pay attention: this method compares strings ONLY. So if you need count by nested properties you should  pass exactly property of type String. 

I mean like this: 

this.restaurants.countBy({ category: { slug } }); 
or 

this.restaurants.count({ where: { category: { slug } } });

type Category = {
name: string,
slug: string,
coverImage: string
.....
}
type Restaurant = {
name: string
owner: string
....
category: Category
}

**Comment by netroy** [comment]
[Here](https://github.com/typeorm/typeorm/pull/9737) is my attempt at adding SQL aggregate functions SUM, AVG, MIN, and MAX to the Repository API.

**Comment by netroy** [comment]
Just FYI: This was released in 0.3.12 🎉

**Comment by mfka** [comment]
For anyone looking for docs: https://orkhan.gitbook.io/typeorm/docs/repository-api#repository-api

**Comment by MattLishmanYara** [comment]
> Just FYI: This was released in 0.3.12 🎉

Is it possible to `sum` multiple fields at a time?

Previously, we would have something like

```js
const query = this.createQueryBuilder('example')
.select('SUM(example.column1)')
.addSelect('SUM(example.column2)')
.getRawMany()
```
But if we had joins for entities, that would mean each one of them needs to be passed to a transformer/DTO which can get messy.

It would be great if we could do something like

```js
myRepository.find({where: {name: 'test'}}).sum('column1').sum('column2'); 
```

Or anything like that to allow us to aggregate columns but still utilise the power of TypeORMs entities and eager loading of relations.

Does the current `sum` method support anything like this or is it just for a single field at a time?
